### PR TITLE
apiserver: add lifecycle signal for preshutdown hook

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Ref https://github.com/kubernetes/kubernetes/issues/108483

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by tkashem** [comment]
ref: https://github.com/kubernetes/kubernetes/pull/110026

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 1]
```diff
@@ -429,7 +429,7 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 //           |                                                       |
 //    ShutdownInitiated (shutdownInitiatedCh)                        |
 //           |                                                       |
-// (ShutdownDelayDuration)                                    (PreShutdownHooks)
+// (ShutdownDelayDuration)                               (PreShutdownHooksStopped)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:432]: The convention I was using was basically that inside `()` I was marking something that is happening.

The signals themselves are marked without () (and if there is something in`()` after the signal, it's in alias in the code.

So I think you need to:
- leave this line as it was (basically signalling that preshutdown hooks are running at this point)
- switch line 424 to:
`PreShutdownHooksStopped (preShutdownHooksHasStoppedCh`

**Review by wojtek-t — COMMENTED** [review]
Just one small nit.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 1 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: fixed

File: staging/src/k8s.io/apiserver/pkg/server/lifecycle_signals.go [hunk 2]
```diff
@@ -143,6 +147,7 @@ func newLifecycleSignals() lifecycleSignals {
 	return lifecycleSignals{
 		ShutdownInitiated:          newNamedChannelWrapper("ShutdownInitiated"),
 		AfterShutdownDelayDuration: newNamedChannelWrapper("AfterShutdownDelayDuration"),
+		PreShutdownHooksStopped:    newNamedChannelWrapper("PreShutdownHooksStopped"),
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/server/lifecycle_signals.go:150]: don't we have a test that should fail now?

https://github.com/kubernetes/kubernetes/blob/f727b5af34dac134aaa5bb5a8f02f78ed6e5291c/staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go#L227-L232

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/lifecycle_signals.go:150]: no, i am not wrapping `PreShutdownHooksStopped` in the test yet, preshutdown hook is still buggy, https://github.com/kubernetes/kubernetes/pull/110026 fixes the bug and the test wraps the hook as well.

**Pull request merged**